README.md
# Baserow

Open source no-code database tool and Airtable alternative. Join our forum on
https://community.baserow.io/ or on Gitter via
https://gitter.im/bramw-baserow/community.

> Warning: From version 0.8.0 to 1.4.1 of Baserow the docker-compose.yml included in 
> the root of this repository has exposed an unsecured redis instance by default on the 
> ip:port 0.0.0.0:6379.
> If you have been running a server which is accessible via the internet using this 
> `docker-compose.yml` then it is possible this redis server has been exposed to the 
> internet depending on your operating system, networking setup and firewall settings.
> Please update to the latest version of Baserow, 1.4.2 which no longer exposes the 
> redis, db and mjml containers. Additionally, 1.4.2 only exposes the backend, 
> web-frontend and media containers on localhost and not on 0.0.0.0. If you were 
> previously relying on `docker-compose.yml` publicly exposing Baserow then please 
> follow the [Make Baserow publicly accessible](./docs/guides/baserow-docker-how-to.md)
> section in the docker how to guide.

[![Deploy to Heroku](https://www.herokucdn.com/deploy/button.svg)](https://heroku.com/deploy?template=https://github.com/bram2w/baserow/tree/master)
*Beta*

**We're hiring** remote developers! More information at 
https://baserow.io/jobs/experienced-full-stack-developer.

![Baserow screenshot](docs/assets/screenshot.png "Baserow screenshot")

## Introduction

Create own database without technical experience. Our user friendly no-code tool gives
you the powers of a developer without leaving your browser.

* Uses popular frameworks like [Django](https://www.djangoproject.com/) and
  [Nuxt.js](https://nuxtjs.org/).
* Uses [PostgreSQL](https://www.postgresql.org/).
* Open source.
* MIT license.
* Self hosted.
* Headless and API first.
* Modular architecture.
* Supports plugins.
* Includes a plugin boilerplate.
* Unlimited rows.
* Runs in your browser.
* Alternative to Airtable.

## Self host

* [**Heroku**: Easily install and scale up on Heroku by using the template.](./docs/guides/installation/install-on-heroku.md)
* [**Cloudron**: Install and update Baserow on your own Cloudron server.](./docs/guides/installation/install-on-cloudron.md)
* [**Ubuntu**: Manually install Baserow step by step on a clean Ubuntu server. **not recommended**](./docs/guides/installation/install-on-ubuntu.md)

## Try out Baserow on your machine 

If you just want to try out Baserow, you can create an account on the SaaS version at
https://baserow.io. Just click on the create account or register button at the 
homepage.

If you want to try out Baserow on your own computer, you can easily start a local 
environment via `docker-compose`. Just clone the repository, run the following commands
and visit http://localhost:3000 in your browser.

```
$ git clone --branch master https://gitlab.com/bramw/baserow.git
$ cd baserow
$ docker-compose up 
```

More detailed instructions and more information about running baserow locally check 
[here](docs/guides/running-baserow-locally.md) or at 
https://baserow.io/docs/guides/running-baserow-locally.

## Development environment

If you want to contribute to Baserow you need to setup the development environment. 
Execute the following commands to start a development version of the baserow environment
which has debug and hot re-loading features enabled.

```
$ git clone --branch master https://gitlab.com/bramw/baserow.git
$ cd baserow
$ docker-compose -f docker-compose.yml -f docker-compose.dev.yml up
$ # OR use our ./dev.sh script which also ensures your dev containers run as your user
$ ./dev.sh --build
```

The Baserow development environment is now running. Visit http://localhost:3000 in your
browser and you should see a working version in development mode.

More detailed instructions and more information about the development environment can 
be found [here](./docs/development/development-environment.md) or at 
https://baserow.io/docs/development/development-environment.

## Plugin development

Because of the modular architecture of Baserow it is possible to create plugins. Make 
your own fields, views, applications, pages or endpoints. We also have a plugin 
boilerplate to get you started right away. More information can be found in the 
[plugin introduction](./docs/plugins/introduction.md) and in the 
[plugin boilerplate docs](./docs/plugins/boilerplate.md).

## Official documentation

The official documentation can be found on the website at https://baserow.io/docs/index
or [here](./docs/index.md) inside the repository. The API docs can be found here at 
https://api.baserow.io/api/redoc/ or if you are looking for the OpenAPI schema here
https://api.baserow.io/api/schema.json.

## Become a sponsor

If you would like to get new features faster, then you might want to consider becoming
a sponsor. By becoming a sponsor we can spend more time on Baserow which means faster
development.

[Become a GitHub Sponsor](https://github.com/sponsors/bram2w)

## Meta

Created by Baserow B.V. - bram@baserow.io.

Distributes under the MIT license. See `LICENSE` for more information.

Version: 1.7.1

The official repository can be found at https://gitlab.com/bramw/baserow.

The changelog can be found [here](./changelog.md).

Become a GitHub Sponsor [here](https://github.com/sponsors/bram2w).

Community chat via https://gitter.im/bramw-baserow/community.


changelog.md
# Changelog

## Released (2021-11-25)

* Fix trashing tables and related link fields causing the field dependency graph to 
  become invalid. 
* Fixed not executing premium tests.

## Released (2021-11-24)

* Fixed a bug where the frontend would fail hard if a table with no views was accessed.
* Tables can now be opened in new browser tabs.
* **Breaking Change**: Baserow's `docker-compose.yml` now allows setting the MEDIA_URL 
  env variable. If using MEDIA_PORT you now need to set MEDIA_URL also.
* **Breaking Change**: Baserow's `docker-compose.yml` container names have changed to
  no longer be hardcoded to prevent naming clashes.
* Added a licensing system for the premium version.
* Fixed bug where it was possible to create duplicate trash entries. 
* Fixed propType validation error when converting from a date field to a boolean field.
* Deprecate internal formula field function field_by_id.
* Made it possible to change user information.
* Added table webhooks functionality.
* Added extra indexes for user tables increasing performance.
* Add lookup field type.
* Add aggregate formula functions and the lookup formula function.
* Fixed date_diff formula function.
* Fixed a bug where the frontend would fail hard when converting a multiple select field
  inside the row edit modal.
* Added the kanban view.
* New templates:
    * House Search
    * Personal Health Log
    * Job Search
    * Single Trip Planner
    * Software Application Bug Tracker
* Updated templates:
    * Commercial Property Management
    * Company Asset Tracker
    * Wedding Planner
    * Blog Post Management
    * Home Inventory
    * Book Writing Guide
    * Political Campaign Contributions
    * Applicant Tracker

## Released (2021-10-05)

* Introduced new endpoint to get and update user account information.
* Fixed bug where a user could not be edited in the admin interface without providing 
  a password.
* Fixed bug where sometimes fields would not be ordered correctly in view exports.
* Fixed bug where brand-new fields weren't included in view exports.
* Fixed error when pasting into a single select field.
* Pasting the value of a single select option into a single select field now selects the
  first option with that value.
* The API now returns appropriate errors when trying to create a field with a name which is too long.
* Importing table data with a column name that is too long will now truncate that name.
* Fixed error when rapidly switching between template tables or views in the template 
  preview.
* Upgraded Django to version 3.2.6 and also upgraded all other backend libraries to 
  their latest versions.
* Fix minor error that could sometimes occur when a row and it's table/group/database
  were deleted in rapid succession.
* Fix accidentally locking of too many rows in various tables during update operations.
* Introduced the has file type filter.
* Fixed bug where the backend would fail hard when an invalid integer was provided as
  'before_id' when moving a row by introducing a decorator to validate query parameters.
* Fixed bug where copying a cell containing a null value resulted in an error.
* Added "Multiple Select" field type.
* Fixed a bug where the currently selected view was not in the viewport of the parent.
* Fixed a bug where views context would not scroll down after a new view has been added.
* New templates:
    * Recipe Book
    * Healthcare Facility Management
    * Bucket List
    * Apartment Hunt
    * Holiday Shopping
    * Email Marketing Campaigns
    * Book Writing Guide
    * Home Inventory
    * Political Campaign Contributions
* Updated templates:
    * Blog Post Management
* Fixed a bug where the backend would fail hard when trying to order by field name without
  using `user_field_names`.
* Added "Formula" field type with 30+ useful functions allowing dynamic per row
  calculations.

## Released (2021-08-11)

* Made it possible to leave a group.
* Changed web-frontend `/api/docs` route into `/api-docs`.
* Bumped the dependencies.
* The internal setting allowing Baserow to run with the user tables in a separate 
  database has been removed entirely to prevent data integrity issues.
* Fixed bug where the currently selected dropdown item is out of view from the dropdown
  window when scrolling with the arrow keys.
* Introduced link row field has row filter.
* Made the form view compatible with importing and exporting.
* Made it possible to use the "F2"-Key to edit a cell without clearing the cell content.
* Added password validation to password reset page.
* Add backup and restore database management commands.
* Dropped the `old_name` column.
* Hide view types that can't be exported in the export modal.
* Relaxed the URL field validator and made it consistent between the backend and 
  web-frontend.
* Fixed nuxt not restarting correctly using the provided Baserow supervisor config file.
* Added steps on how to configure Baserow to send emails in the install-on-ubuntu guide.
* Enabled password validation in the backend.
* **Premium**: You can now comment and discuss rows with others in your group, click the
  expand row button at the start of the row to view and add comments.
* Added "Last Modified" and "Created On" field types.
* Fixed moment issue if core is installed as a dependency.
* New templates:
  * Blog Post Management
* Updated templates:
  * Personal Task Manager
  * Wedding Planning
  * Book Catalog
  * Applicant Tracker
  * Project Tracker

## Released (2021-07-16)

* Fix bug preventing fields not being able to be converted to link row fields in some
  situations.

## Released (2021-07-15)

* **Breaking Change**: Baserow's `docker-compose.yml` no longer exposes ports for 
  the `db`, `mjml` and `redis` containers for security reasons. 
* **Breaking Change**: `docker-compose.yml` will by default only expose Baserow on 

backend/src/baserow/config/settings/base.py
import datetime
import os
from urllib.parse import urlparse, urljoin

from corsheaders.defaults import default_headers

BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))

# SECURITY WARNING: keep the secret key used in production secret!
SECRET_KEY = os.getenv("SECRET_KEY", "CHANGE_THIS_TO_SOMETHING_SECRET_IN_PRODUCTION")

# SECURITY WARNING: don't run with debug turned on in production!
DEBUG = False

ALLOWED_HOSTS = ["localhost"]

INSTALLED_APPS = [
    "django.contrib.auth",
    "django.contrib.contenttypes",
    "django.contrib.sessions",
    "django.contrib.messages",
    "django.contrib.staticfiles",
    "rest_framework",
    "corsheaders",
    "channels",
    "mjml",
    "drf_spectacular",
    "djcelery_email",
    "baserow.core",
    "baserow.api",
    "baserow.ws",
    "baserow.contrib.database",
    "baserow_premium",
]

ADDITIONAL_APPS = os.getenv("ADDITIONAL_APPS", None)
if ADDITIONAL_APPS is not None:
    INSTALLED_APPS += ADDITIONAL_APPS.split(",")

MIDDLEWARE = [
    "corsheaders.middleware.CorsMiddleware",
    "django.middleware.security.SecurityMiddleware",
    "django.contrib.sessions.middleware.SessionMiddleware",
    "django.middleware.common.CommonMiddleware",
    "django.middleware.csrf.CsrfViewMiddleware",
    "django.contrib.auth.middleware.AuthenticationMiddleware",
    "django.contrib.messages.middleware.MessageMiddleware",
    "django.middleware.clickjacking.XFrameOptionsMiddleware",
]

ROOT_URLCONF = "baserow.config.urls"

TEMPLATES = [
    {
        "BACKEND": "django.template.backends.django.DjangoTemplates",
        "DIRS": [],
        "APP_DIRS": True,
        "OPTIONS": {
            "context_processors": [
                "django.template.context_processors.debug",
                "django.template.context_processors.request",
                "django.contrib.auth.context_processors.auth",
                "django.contrib.messages.context_processors.messages",
            ],
        },
    },
]

WSGI_APPLICATION = "baserow.config.wsgi.application"
ASGI_APPLICATION = "baserow.config.asgi.application"

REDIS_HOST = os.getenv("REDIS_HOST", "redis")
REDIS_PORT = os.getenv("REDIS_PORT", "6379")
REDIS_USERNAME = os.getenv("REDIS_USER", "")
REDIS_PASSWORD = os.getenv("REDIS_PASSWORD", "")
REDIS_PROTOCOL = os.getenv("REDIS_PROTOCOL", "redis")
REDIS_URL = os.getenv(
    "REDIS_URL",
    f"{REDIS_PROTOCOL}://{REDIS_USERNAME}:{REDIS_PASSWORD}@{REDIS_HOST}:{REDIS_PORT}/0",
)

CELERY_BROKER_URL = REDIS_URL
CELERY_TASK_ROUTES = {
    "baserow.contrib.database.export.tasks.run_export_job": {"queue": "export"},
    "baserow.contrib.database.export.tasks.clean_up_old_jobs": {"queue": "export"},
    "baserow.core.trash.tasks.mark_old_trash_for_permanent_deletion": {
        "queue": "export"
    },
    "baserow.core.trash.tasks.permanently_delete_marked_trash": {"queue": "export"},
}
CELERY_SOFT_TIME_LIMIT = 60 * 5
CELERY_TIME_LIMIT = CELERY_SOFT_TIME_LIMIT + 60

CELERY_REDBEAT_REDIS_URL = REDIS_URL
# Explicitly set the same value as the default loop interval here so we can use it
# later. CELERY_BEAT_MAX_LOOP_INTERVAL < CELERY_REDBEAT_LOCK_TIMEOUT must be kept true
# as otherwise a beat instance will acquire the lock, do scheduling, go to sleep for
# CELERY_BEAT_MAX_LOOP_INTERVAL before waking up where it assumes it still owns the lock
# however if the lock timeout is less than the interval the lock will have been released
# and the beat instance will crash as it attempts to extend the lock which it no longer
# owns.
CELERY_BEAT_MAX_LOOP_INTERVAL = os.getenv("CELERY_BEAT_MAX_LOOP_INTERVAL", 20)
# By default CELERY_REDBEAT_LOCK_TIMEOUT = 5 * CELERY_BEAT_MAX_LOOP_INTERVAL
# Only one beat instance can hold this lock and schedule tasks at any one time.
# This means if one celery-beat instance crashes any other replicas waiting to take over
# will by default wait 25 minutes until the lock times out and they can acquire
# the lock to start scheduling tasks again.
# Instead we just set it to be slightly longer than the loop interval that beat uses.
# This means beat wakes up, checks the schedule and extends the lock every
# CELERY_BEAT_MAX_LOOP_INTERVAL seconds. If it crashes or fails to wake up
# then 80 seconds after the lock was last extended it will be released and a new
# scheduler will acquire the lock and take over.
CELERY_REDBEAT_LOCK_TIMEOUT = os.getenv(
    "CELERY_REDBEAT_LOCK_TIMEOUT", CELERY_BEAT_MAX_LOOP_INTERVAL + 60
)

CHANNEL_LAYERS = {
    "default": {
        "BACKEND": "channels_redis.core.RedisChannelLayer",
        "CONFIG": {
            "hosts": [REDIS_URL],
        },
    },
}

# Database
# https://docs.djangoproject.com/en/2.2/ref/settings/#databases

DATABASES = {
    "default": {
        "ENGINE": "django.db.backends.postgresql",
        "NAME": os.getenv("DATABASE_NAME", "baserow"),
        "USER": os.getenv("DATABASE_USER", "baserow"),
        "PASSWORD": os.getenv("DATABASE_PASSWORD", "baserow"),
        "HOST": os.getenv("DATABASE_HOST", "db"),
        "PORT": os.getenv("DATABASE_PORT", "5432"),
    }
}

# Should contain the database connection name of the database where the user tables
# are stored. This can be different than the default database because there are not
# going to be any relations between the application schema and the user schema.
USER_TABLE_DATABASE = "default"

# Password validation
# https://docs.djangoproject.com/en/2.2/ref/settings/#auth-password-validators

AUTH_PASSWORD_VALIDATORS = [
    {
        "NAME": "django.contrib.auth.password_validation.MinimumLengthValidator",
    },
    {
        "NAME": "baserow.core.user.password_validation.MaximumLengthValidator",
    },
]

# We need the `AllowAllUsersModelBackend` in order to respond with a proper error
# message when the user is not active. The only thing it does, is allowing non active
# users to authenticate, but the user still can't obtain or use a JWT token or database
# token because the user needs to be active to use that.
AUTHENTICATION_BACKENDS = ["django.contrib.auth.backends.AllowAllUsersModelBackend"]

# Internationalization
# https://docs.djangoproject.com/en/2.2/topics/i18n/

LANGUAGE_CODE = "en"

LANGUAGES = [
    ("en", "English"),
    ("fr", "French"),
]

TIME_ZONE = "UTC"

USE_I18N = True

USE_L10N = True

USE_TZ = True

# Static files (CSS, JavaScript, Images)
# https://docs.djangoproject.com/en/2.2/howto/static-files/

STATIC_URL = "/static/"

REST_FRAMEWORK = {
    "DEFAULT_PERMISSION_CLASSES": ("rest_framework.permissions.IsAuthenticated",),
    "DEFAULT_AUTHENTICATION_CLASSES": (
        "baserow.api.authentication.JSONWebTokenAuthentication",
    ),
    "DEFAULT_RENDERER_CLASSES": ("rest_framework.renderers.JSONRenderer",),
    "DEFAULT_SCHEMA_CLASS": "baserow.api.openapi.AutoSchema",
}

CORS_ORIGIN_ALLOW_ALL = True
CORS_ALLOW_HEADERS = list(default_headers) + [
    "WebSocketId",
]

# JWT设置
JWT_AUTH = {
    "JWT_EXPIRATION_DELTA": datetime.timedelta(seconds=60 * 60),
    "JWT_ALLOW_REFRESH": True,
    "JWT_REFRESH_EXPIRATION_DELTA": datetime.timedelta(days=7),
    "JWT_AUTH_HEADER_PREFIX": "JWT",
    "JWT_RESPONSE_PAYLOAD_HANDLER": "baserow.api.user.jwt.jwt_response_payload_handler",
}

# oauth第三方登陆配置
FELO_CODE_URL = 'https://passport.felo.me/oauth/authorize?'  # 请求code地址
FELO_TOKEN_URL = 'https://passport.felo.me/oauth/token'  # 通过code请求token地址
FELO_INFO_URL = 'https://open.felo.me/sdk/user/current'  # 通过token获取用户信息
FELO_CLIENT_ID = 'ydFC1EF960E67640199449EB2C6A6F3921'  # 应用ID
FELO_CLIENT_SECRET = 'JKxWwd9Q+8DMP0IFtam8JHDPVzdGvw0HWjRPdD42fSM='  # 应用密钥
FELO_REDIRECT_URI = 'http://localhost:3000/oauthLogin'  # 授权后携带授权码重定向到的网页地址（urlencode格式）

SPECTACULAR_SETTINGS = {
    "TITLE": "Baserow API spec",
    "DESCRIPTION": "",
    "CONTACT": {"url": "https://baserow.io/contact"},
    "LICENSE": {
        "name": "MIT",
        "url": "https://gitlab.com/bramw/baserow/-/blob/master/LICENSE",
    },
    "VERSION": "1.7.1",
    "SERVE_INCLUDE_SCHEMA": False,
    "TAGS": [
        {"name": "Settings"},
        {"name": "User"},
        {"name": "User files"},
        {"name": "Groups"},
        {"name": "Group invitations"},
        {"name": "Templates"},
        {"name": "Trash"},
        {"name": "Applications"},
        {"name": "Database tables"},
        {"name": "Database table fields"},
        {"name": "Database table views"},
        {"name": "Database table view filters"},
        {"name": "Database table view sortings"},
        {"name": "Database table grid view"},
        {"name": "Database table form view"},
        {"name": "Database table kanban view"},
        {"name": "Database table rows"},
        {"name": "Database table export"},
        {"name": "Database table webhooks"},
        {"name": "Database tokens"},
        {"name": "Admin"},
    ],
    "ENUM_NAME_OVERRIDES": {
        "NumberDecimalPlacesB02Enum": [
            (0, "1"),
            (1, "1.0"),
            (2, "1.00"),
            (3, "1.000"),
            (4, "1.0000"),
            (5, "1.00000"),
        ],
        "NumberDecimalPlaces0c0Enum": [
            (1, "1.0"),
            (2, "1.00"),
            (3, "1.000"),
            (4, "1.0000"),
            (5, "1.00000"),
        ],
    },
}

# The storage must always overwrite existing files.
DEFAULT_FILE_STORAGE = "baserow.core.storage.OverwriteFileSystemStorage"

MJML_BACKEND_MODE = "tcpserver"
MJML_TCPSERVERS = [
    (os.getenv("MJML_SERVER_HOST", "mjml"), int(os.getenv("MJML_SERVER_PORT", 28101))),
]

PUBLIC_BACKEND_URL = os.getenv("PUBLIC_BACKEND_URL", "http://localhost:8000")
PUBLIC_WEB_FRONTEND_URL = os.getenv("PUBLIC_WEB_FRONTEND_URL", "http://localhost:3000")
PRIVATE_BACKEND_URL = os.getenv("PRIVATE_BACKEND_URL", "http://backend:8000")
PUBLIC_BACKEND_HOSTNAME = urlparse(PUBLIC_BACKEND_URL).hostname
PUBLIC_WEB_FRONTEND_HOSTNAME = urlparse(PUBLIC_WEB_FRONTEND_URL).hostname
PRIVATE_BACKEND_HOSTNAME = urlparse(PRIVATE_BACKEND_URL).hostname

if PUBLIC_BACKEND_HOSTNAME:
    ALLOWED_HOSTS.append(PUBLIC_BACKEND_HOSTNAME)

if PRIVATE_BACKEND_HOSTNAME:
    ALLOWED_HOSTS.append(PRIVATE_BACKEND_HOSTNAME)

FROM_EMAIL = os.getenv("FROM_EMAIL", "no-reply@localhost")
RESET_PASSWORD_TOKEN_MAX_AGE = 60 * 60 * 48  # 48 hours
ROW_PAGE_SIZE_LIMIT = 200  # How many rows can be requested at once.
TRASH_PAGE_SIZE_LIMIT = 200  # How many trash entries can be requested at once.
ROW_COMMENT_PAGE_SIZE_LIMIT = 200  # How many row comments can be requested at once.

# The amount of rows that can be imported when creating a table.
INITIAL_TABLE_DATA_LIMIT = None
if "INITIAL_TABLE_DATA_LIMIT" in os.environ:
    INITIAL_TABLE_DATA_LIMIT = int(os.getenv("INITIAL_TABLE_DATA_LIMIT"))

MEDIA_URL_PATH = "/media/"
MEDIA_URL = os.getenv("MEDIA_URL", urljoin(PUBLIC_BACKEND_URL, MEDIA_URL_PATH))
MEDIA_ROOT = os.getenv("MEDIA_ROOT", "/baserow/media")

# Indicates the directory where the user files and user thumbnails are stored.
USER_FILES_DIRECTORY = "user_files"
USER_THUMBNAILS_DIRECTORY = "thumbnails"
USER_FILE_SIZE_LIMIT = 1024 * 1024 * 20  # 20MB

EXPORT_FILES_DIRECTORY = "export_files"
EXPORT_CLEANUP_INTERVAL_MINUTES = 5
EXPORT_FILE_EXPIRE_MINUTES = 60

EMAIL_BACKEND = "djcelery_email.backends.CeleryEmailBackend"

if os.getenv("EMAIL_SMTP", ""):
    CELERY_EMAIL_BACKEND = "django.core.mail.backends.smtp.EmailBackend"
    # EMAIL_SMTP_USE_TLS OR EMAIL_SMTP_USE_TLS for backwards compatibility after
    # fixing #448.
    EMAIL_USE_TLS = bool(os.getenv("EMAIL_SMTP_USE_TLS", "")) or bool(
        os.getenv("EMAIL_SMPT_USE_TLS", "")
    )
    EMAIL_HOST = os.getenv("EMAIL_SMTP_HOST", "localhost")
    EMAIL_PORT = os.getenv("EMAIL_SMTP_PORT", "25")
    EMAIL_HOST_USER = os.getenv("EMAIL_SMTP_USER", "")
    EMAIL_HOST_PASSWORD = os.getenv("EMAIL_SMTP_PASSWORD", "")
else:
    CELERY_EMAIL_BACKEND = "django.core.mail.backends.console.EmailBackend"

# Configurable thumbnails that are going to be generated when a user uploads an image
# file.
USER_THUMBNAILS = {"tiny": [None, 21], "small": [48, 48]}

# The directory that contains the all the templates in JSON format. When for example
# the `sync_templates` management command is called, then the templates in the
# database will be synced with these files.
APPLICATION_TEMPLATES_DIR = os.path.join(BASE_DIR, "../../../templates")
# The template that must be selected when the user first opens the templates select
# modal.
DEFAULT_APPLICATION_TEMPLATE = "project-tracker"

MAX_FIELD_LIMIT = 1500

# If you change this default please also update the default for the web-frontend found
# in web-frontend/modules/core/module.js:55
HOURS_UNTIL_TRASH_PERMANENTLY_DELETED = os.getenv(
    "HOURS_UNTIL_TRASH_PERMANENTLY_DELETED", 24 * 3
)
OLD_TRASH_CLEANUP_CHECK_INTERVAL_MINUTES = 5

MAX_ROW_COMMENT_LENGTH = 10000

DEFAULT_AUTO_FIELD = "django.db.models.AutoField"

# For now force the old os dependant behaviour of file uploads as users might be relying
# on it. See
# https://docs.djangoproject.com/en/3.2/releases/3.0/#new-default-value-for-the-file-upload-permissions-setting
FILE_UPLOAD_PERMISSIONS = None

MAX_FORMULA_STRING_LENGTH = 10000
MAX_FIELD_REFERENCE_DEPTH = 1000
UPDATE_FORMULAS_AFTER_MIGRATION = bool(
    os.getenv("UPDATE_FORMULAS_AFTER_MIGRATION", "yes")
)

WEBHOOKS_MAX_CONSECUTIVE_TRIGGER_FAILURES = 8
WEBHOOKS_MAX_RETRIES_PER_CALL = 8
WEBHOOKS_MAX_PER_TABLE = 20
WEBHOOKS_MAX_CALL_LOG_ENTRIES = 10
WEBHOOKS_REQUEST_TIMEOUT_SECONDS = 5
